### PR TITLE
leland.cpp cleanups

### PR DESCRIPTION
**Pull request by cam900**
leland.cpp : Cleanup duplicates, Add generic_latch_16_device for maincpu to soundcpu comms, Improve tags, Reduce runtime tag lookups, Make drawing function relate to cliprect

### CODE REVIEW

File: src/mame/audio/leland.cpp [hunk 1]
```diff
@@ -471,23 +462,15 @@ WRITE8_MEMBER( leland_80186_sound_device::leland_80186_control_w )
...
-	machine().scheduler().synchronize(timer_expired_delegate(FUNC(leland_80186_sound_device::command_lo_sync), this), data);
+	m_soundlatch->write(space, offset, data & 0xff, mem_mask & 0xff);
 }
 
 
 WRITE8_MEMBER( leland_80186_sound_device::leland_80186_command_hi_w )
 {
-	if (LOG_COMM) logerror("%s:Write sound command latch hi = %02X\n", machine().describe_context(), data);
-	m_sound_command = (m_sound_command & 0x00ff) | (data << 8);
+	m_soundlatch->write(space, offset, data << 8, mem_mask << 8);
```
Comment (cuavas) [inline src/mame/audio/leland.cpp]: This won't actually work - the 16-bit latch device doesn't honour the mask.  You need to check that your changes actually build and work properly.

**Comment by cam900** [comment]
It's compilable and tested, Most of games are working without any issues.

**Comment by cuavas** [comment]
When you say "most games are working" do you mean you tested most games, and all the ones you tested are working, or do you mean that you tested all the games, and most of them worked but others had issues?

**Comment by Tafoid** [comment]
Some testing indicates a severe performance issues with a portion of machines in the driver (listed below,  in many cases over 100x slower).  

- alleymas - leland.cpp
Average speed: 10.83% (2 seconds)
- basebal2 - leland.cpp
Average speed: 11.26% (2 seconds)
- dangerz - leland.cpp
Average speed: 10.99% (2 seconds)
- dblplay - leland.cpp
Average speed: 10.98% (2 seconds)
- mayhem - leland.cpp
Average speed: 6.61% (2 seconds)
- powrplay - leland.cpp
Average speed: 13.48% (2 seconds)
- strkzone - leland.cpp
Average speed: 11.45% (2 seconds)
- upyoural - leland.cpp
Average speed: 10.48% (2 seconds)
- wseries - leland.cpp
Average speed: 11.31% (2 seconds)
- wseries0 - leland.cpp
Average speed: 11.32% (2 seconds)

Problem appears to be in tilemap area, based on **-profile 3 -bench 3** run:

> mame64 wseries -bench 3 -profile 3
>- 86.8% -  26789 : 0000000003ac6120 (tilemap_t::tile_update(unsigned int, unsigned int, unsigned int))
>              0000000003ac6626 (tilemap_t::pixmap_update()+0x0106)
>               0000000000600900 (leland_state::screen_update(screen_device&, bitmap_ind16&, rectangle const&)+0x0030)
>
>- 9.4% -   2911 : 00000000006007e0 (leland_state::leland_get_tile_info(tilemap_t&, tile_data&, unsigned int))
>                 0000000003ac6500 (tilemap_t::tile_update(unsigned int, unsigned int, unsigned int)+0x03e0)
>                 0000000003ac6626 (tilemap_t::pixmap_update()+0x0106)
>
>- 1.1% -    347 : 0000000003ac6520 (tilemap_t::pixmap_update())
>                 0000000000600900 (leland_state::screen_update(screen_device&, bitmap_ind16&, rectangle const&)+0x0030)
>                 0000000003ab6ef4 (screen_device::update_partial(int)+0x0144)




The rest in the driver seem to show a performance increase in most cases 25% or more.
Examples:
 - asylum - leland.cpp
Before: Average speed: 772.01% (2 seconds)
After: Average speed: 990.51% (2 seconds)
- ataxx - leland.cpp
Before: Average speed: 846.73% (2 seconds)
After: Average speed: 1242.53% (2 seconds)

**Comment by cuavas** [comment]
Performance is back with a3b446262d761e05707f8779d5a819a31d37987f - it's easy to screw yourself over with tilemaps, so be careful.

**Pull request merged**